cache/eventsource.go

```diff
@@ -0,0 +1,77 @@
+package cache
+
+import (
+	"context"
+	"sync/atomic"
+	"time"
+
+	rpctypes "go.etcd.io/etcd/api/v3/v3rpc/rpctypes"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+type eventSource struct {
+	cli       *clientv3.Client
+	ctx       context.Context
+	cancel    context.CancelFunc
+	ch        <-chan clientv3.WatchResponse
+	currentRV atomic.Int64
+}
+
+func newEventSource(cli *clientv3.Client, startRev int64, parent context.Context) *eventSource {
+	ctx, cancel := context.WithCancel(parent)
+	es := &eventSource{cli: cli, ctx: ctx, cancel: cancel}
+	es.start(startRev)
+	return es
+}
+
+func (es *eventSource) start(rev int64) {
```

Comment:
Little lost which methods are for intended to be called from caller and which should be called only by the structure itself. Recommend capitalize methods that should be called by external caller. As the structure itself is private you don't need to worry about this being public.